<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Шаг-1.-Обзор-и-подготовка-данных" data-toc-modified-id="Шаг-1.-Обзор-и-подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Обзор и подготовка данных</a></span></li><li><span><a href="#Шаг-2.-Подготовка-к-обучению:-обучающая,-валидационная-и-тестовая-выборки,-признаки" data-toc-modified-id="Шаг-2.-Подготовка-к-обучению:-обучающая,-валидационная-и-тестовая-выборки,-признаки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 2. Подготовка к обучению: обучающая, валидационная и тестовая выборки, признаки</a></span></li><li><span><a href="#Шаг-3.-Исследование-качества-разных-моделей-с-различными-вариантами-гиперпараметров" data-toc-modified-id="Шаг-3.-Исследование-качества-разных-моделей-с-различными-вариантами-гиперпараметров-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Шаг 3. Исследование качества разных моделей с различными вариантами гиперпараметров</a></span></li><li><span><a href="#Шаг-4.-Проверка-качества-модели-на-тестовой-выборке" data-toc-modified-id="Шаг-4.-Проверка-качества-модели-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Шаг 4. Проверка качества модели на тестовой выборке</a></span></li><li><span><a href="#Шаг-5.-Проверка-моделей-на-вменяемость-(sanity-check)" data-toc-modified-id="Шаг-5.-Проверка-моделей-на-вменяемость-(sanity-check)-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Шаг 5. Проверка моделей на вменяемость (sanity check)</a></span></li><li><span><a href="#Шаг-6.-Общий-вывод" data-toc-modified-id="Шаг-6.-Общий-вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Шаг 6. Общий вывод</a></span></li></ul></div>

# Проект 5 спринт

**Цель проекта**

Проанализировать данные по поведению клиентов, пользующихся тарифными планами "Смарт" и "Ультра", и на основании анализа построить модель для задачи классификации с долей правильных ответов >= 0,75, которая сможет рекомендовать подходящий под поведение клиента тариф. 


**Данные для анализа**

На анализ поступила предобработанная выборка из проекта курса «Статистический анализ данных» с данными о поведении клиентов, которые уже перешли на тарифы "Смарт" или "Ультра".

**Шаги (план) проекта**

1.  Шаг 1. Обзор и подготовка данных
2.  Шаг 2. Подготовка к обучению: обучающая, валидационная и тестовая выборки, признаки
3.  Шаг 3. Исследование качества разных моделей с различными вариантами гиперпараметров
4.  Шаг 4. Проверка качества модели на тестовой выборке
5.  Шаг 5. Проверка моделей на вменяемость (sanity check)
6.  Шаг 6. Общий вывод

## Шаг 1. Обзор и подготовка данных

Импортируем необходимые библиотеки для работы с данными в текущем проекте.

In [1]:
import pandas as pd # импорт библиотеки pandas
import numpy as np # импорт библиотеки numpy
from sklearn.model_selection import train_test_split # импорт функции train_test_split из библиотеки sklearn для деления выборки
from sklearn.metrics import accuracy_score # импорт функции для расчета точности прогнозирования
from sklearn.tree import DecisionTreeClassifier # импорт алгоритма "дерево решений" из библиотеки sklearn
from sklearn.ensemble import RandomForestClassifier # импорт алгоритма "случайный лес" из библиотеки sklearn
from sklearn.linear_model import LogisticRegression # импорт алгоритма "логистическая регрессия" из библиотеки sklearn

Откроем таблицу с данными, выведем первые 10 строк.

In [2]:
# чтение файла с данными и сохранение его в переменную df 
df = pd.read_csv('/datasets/users_behavior.csv')
# вывод первых 10 строк
df.head(10)

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
5   58.0   344.56      21.0  15823.37         0
6   57.0   431.64      20.0   3738.90         1
7   15.0   132.40       6.0  21911.60         0
8    7.0    43.39       3.0   2538.67         1
9   90.0   665.41      38.0  17358.61         0

Изучим общую информацию о данных.

In [3]:
df.info() # получение общей информации о структуре датафрейма

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Таблица содержит 5 столбцов, 3214 строк. Наименования столбцов оформлены "змеиным стилем" в нижнем регистре, переименование не требуется.

В столбцах содержатся данные двух типов:

* 4 столбца содержат числа с плавающей запятой float64

* 1 столбец содержит целые числа int64

Пустых значений в столбцах нет, следовательно, обработка пропусков не требуется.

Изменим тип данных в столбце calls на int и все типы 64 на типы 32 для экономии памяти и проверим данные на полные дубликаты.

**Изменение типа данных в столбцах**

In [4]:
# замена типа данных на целое число
df['calls'] = df['calls'].astype('int')

In [5]:
# замена типа данных float64 на float32
df[df.select_dtypes(np.float64).columns] = df.select_dtypes(np.float64).astype(np.float32)

In [6]:
# замена типа данных int64 на int32
df[df.select_dtypes(np.int64).columns] = df.select_dtypes(np.int64).astype(np.int32)

Проверим корректность изменения типов данных.

In [7]:
df.info() # получение общей информации о структуре датафрейма

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   int32  
 1   minutes   3214 non-null   float32
 2   messages  3214 non-null   float32
 3   mb_used   3214 non-null   float32
 4   is_ultra  3214 non-null   int32  
dtypes: float32(3), int32(2)
memory usage: 62.9 KB


Типы данных изменены корректно.

**Проверка данных на наличие явных дубликатов**

In [8]:
# подстчет строк - явных дубликатов
df.duplicated().sum()

0

Явных дубликатов в данных нет.

<div style="border:solid blue 2px; padding: 20px"> 

**Вывод**  

Таблица открыта, данные адекватны.

Проведена проверка для определения необходимости дополнительной предобработки данных:
    
1. Проведена замена типов данных в колонках calls, minutes, messages, mb_used, is_ultra. 

2. Переименование колонок не требуется.
    
3. Пропусков и дубликатов в данных нет. 

</div>

## Шаг 2. Подготовка к обучению: обучающая, валидационная и тестовая выборки, признаки

Поскольку в данном прокте у нас нет дополнительной тестовой выборки, разделим исходные данные на обучающую, валидационную и тестовую выборки в соотношении 3:1:1.

In [9]:
# выделение 40% данных датасета для формирования валидационной и тестовой выборок
df_train, df_tv = train_test_split(df, test_size=0.40, random_state=12345)

In [10]:
# разделение выделенных 40% данных пополам для формирования валидационной и тестовой выборок (по 20% от всего датасета)
df_valid, df_test = train_test_split(df_tv, test_size=0.50, random_state=12345)

Оценим размер получившихся выборок.

In [11]:
# вывод размера выборки
df_train.shape

(1928, 5)

In [12]:
# вывод размера выборки
df_valid.shape

(643, 5)

In [13]:
# вывод размера выборки
df_test.shape

(643, 5)

Создадим переменные с признаками и целевым признаком для всех трех типов выборок.

**Обучающая выборка**

In [14]:
# создание переменных с признаками (features) и целевым признаком (target)
features_train = df_train.drop(['is_ultra'], axis=1)
target_train = df_train['is_ultra']

**Валидационная выборка**

In [15]:
# создание переменных с признаками (features) и целевым признаком (target)
features_valid = df_valid.drop(['is_ultra'], axis=1)
target_valid = df_valid['is_ultra']

**Тестовая выборка**

In [16]:
# создание переменных с признаками (features) и целевым признаком (target)
features_test = df_test.drop(['is_ultra'], axis=1)
target_test = df_test['is_ultra']

<div style="border:solid blue 2px; padding: 20px"> 

**Вывод**  

Данные успешно разделены на выборки: обучающую (60% от всего датасета), валидационную (20%) и тестовую (20%).  
    
Для всех трех типов выборок заданы переменные с признаками и целевым признаком.    
 
 
</div>

## Шаг 3. Исследование качества разных моделей с различными вариантами гиперпараметров

Поскольку наш целевой признак - категориальный, и имеет только 2 варианта (0 и 1), нам необходимо решить задачу бинарной (двоичной) классификации. Используем наиболее популярные модели, подходящие для решения такого рода задач.

**Модель Дерево решений**

Создадим цикл, который создаст и обучит модель на данных обучающей выборки, перебрав различные варианты глубины дерева решений от 1 до 10, и на основе максимального значения accuracy на данных валидационной выборки определит модель с оптимальной глубиной дерева.

In [17]:
# создание цикла для определения лучшей модели
best_model = None
best_result = 0
best_depth = 0
for depth in range(1, 11):
    model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_tree.fit(features_train, target_train)
    predictions_tree_valid = model_tree.predict(features_valid)
    result_tree_valid = accuracy_score(target_valid, predictions_tree_valid)
    if result_tree_valid > best_result:
        best_model = model_tree
        best_result = result_tree_valid
        best_depth = depth
        
print("Accuracy наилучшей модели на валидационной выборке:", best_result, "Глубина дерева:", best_depth)

Accuracy наилучшей модели на валидационной выборке: 0.7853810264385692 Глубина дерева: 3


**Модель Случайный лес**

Создадим цикл, который создаст и обучит модель на данных обучающей выборки, перебрав различные варианты количества оценщиков от 1 до 20, и на основе максимального значения accuracy на данных валидационной выборки определит модель с оптимальным количеством оценщиков.

In [18]:
# создание цикла для определения лучшей модели
best_model = None
best_result = 0
best_est = 0
for est in range(1, 21):
    model_forest = RandomForestClassifier(random_state=12345, n_estimators=est)
    model_forest.fit(features_train, target_train)
    predictions_forest_valid = model_forest.predict(features_valid)
    result_forest_valid = accuracy_score(target_valid, predictions_forest_valid)
    if result_forest_valid > best_result:
        best_model = model_forest
        best_result = result_forest_valid
        best_est = est

print("Accuracy наилучшей модели на валидационной выборке:", best_result, "Количество оценщиков:", best_est)

Accuracy наилучшей модели на валидационной выборке: 0.7931570762052877 Количество оценщиков: 18


**Модель логистической регрессии**

Создадим модель логистической регрессии и оценим точность ее предсказания.

In [19]:
# создание модели
model_regression = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)

In [20]:
# обучение модели на обучающей выборке и оценка точности предсказания на валидационной выборке
model_regression.fit(features_train, target_train)
predictions_regression_valid = model_regression.predict(features_valid)
result_regression_valid = accuracy_score(target_valid, predictions_regression_valid)
print("Accuracy модели на валидационной выборке:", result_regression_valid)

Accuracy модели на валидационной выборке: 0.7107309486780715


<div style="border:solid blue 2px; padding: 20px"> 

**Вывод**  

Для выбора оптимальной модели обучения были протестированы три модели:
    
* решающее дерево, 
* случайный лес,
* логистическая регрессия.     
 
Все три модели были обучены на обучающей выборке, после чего была оценена точность их предсказаний на валидационной выборке.

Модели показали следующие результаты:
    
* решающее дерево - Accuracy наилучшей модели на валидационной выборке: 0.7853810264385692 Глубина дерева: 3, 
* случайный лес - Accuracy наилучшей модели на валидационной выборке: 0.7931570762052877 Количество оценщиков: 18
* логистическая регрессия -  Accuracy модели на валидационной выборке: 0.7107309486780715 
    
Наилучший результат у модели "Случайный лес" при 18 оценщиках (Accuracy = 0.7931570762052877).    
 
</div>

**Полнота, точность, F1-мера**

Оценим метрики качества модели с лучшим значением accuracy "Случайный лес" при 18 оценщиках.

In [21]:
# импорт необходимых функций
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import f1_score

# прогнозирование модели и оценка метрик качества сделанного ею прогноза
model = RandomForestClassifier(random_state=12345, n_estimators=18)
model.fit(features_train, target_train)
predicted_train = model.predict(features_train)
print('Полнота:', recall_score(target_train, predicted_train))
print('Точность:', precision_score(target_train, predicted_train))
print('F1-мера:',f1_score(target_train, predicted_train))

Полнота: 0.9629005059021922
Точность: 0.9965095986038395
F1-мера: 0.9794168096054889


Все метрики достаточно высокие, стремящиеся к 1, что подтверждает корректность выбора модели.

## Шаг 4. Проверка качества модели на тестовой выборке

Проверим точность предсказания моделей на данных тестовой выборки.

Задача - выбрать наилучшую модель с долей правильных ответов >= 0.75.

**Модель Дерево решений**

Оценим точность предсказаний на тестовой выборке для модели "Решающее дерево" с глубиной дерева 3.

In [22]:
# оценка точности предсказаний
model_tree = DecisionTreeClassifier(random_state=12345, max_depth=3)
model_tree.fit(features_train, target_train)
predictions_tree_test = model_tree.predict(features_test)
result_tree_test = accuracy_score(target_test, predictions_tree_test)
print("Accuracy модели на тестовой выборке:", result_tree_test)

Accuracy модели на тестовой выборке: 0.7791601866251944


Результат точности предсказаний выше 0,75, но ниже, чем на валидационной выборке. Имеет место некоторое переобучение модели.

**Модель Случайный лес**

Оценим точность предсказаний на тестовой выборке для модели "Случайный лес" с 18 оценщиками.

In [23]:
# оценка точности предсказаний
model_forest = RandomForestClassifier(random_state=12345, n_estimators=18)
model_forest.fit(features_train, target_train)
predictions_forest_test = model_forest.predict(features_test)
result_forest_test = accuracy_score(target_test, predictions_forest_test)
print("Accuracy модели на тестовой выборке:", result_forest_test)  

Accuracy модели на тестовой выборке: 0.7853810264385692


Результат точности предсказаний выше 0,75, но ниже, чем на валидационной выборке. Имеет место некоторое переобучение модели.

**Модель логистической регрессии**

Оценим точность предсказаний на тестовой выборке для модели логистической регрессии.

In [24]:
# оценка точности предсказаний
model_regression = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)
model_regression.fit(features_train, target_train)
predictions_regression_test = model_regression.predict(features_test)
result_regression_test = accuracy_score(target_test, predictions_regression_test)
print("Accuracy модели на тестовой выборке:", result_regression_test)

Accuracy модели на тестовой выборке: 0.6842923794712286


Результат неудовлетворительный, значительно ниже, чем на валидационной выборке, и ниже необходимого значения 0,75. Очевидно, что модель переобучена.

<div style="border:solid blue 2px; padding: 20px"> 

**Вывод**  

Проверка точности предсказаний моделей на тестовой выборке показала, что точность предсказаний всех моделей на тестовой выборке снизилась по сравнению с предсказаниями на валидационной выборке, что с большой долей вероятности говорит о переобучении моделей.


И на валидационной, и на тестовой выборке наилучшие результаты показала модель Случайный лес с 18 оценщиками (на валидационной выборке accuracy = 0.793, на тестовой выборке accuracy = 0.785).
    
Сводные результаты эксперимента:    

|Модель                     |Точность на валидационной выборке |Точность на тестовой выборке |
|:-------------------------:|---------------------------------:|----------------------------:|
|Решающее дерево, глубина 3 |0.7853810264385692                |0.7791601866251944           |
|Случайный лес, оценщиков 18|0.7931570762052877                |0.7853810264385692           | 
|Логистическая регрессия    |0.7107309486780715                |0.6842923794712286           |
 
</div>

## Шаг 5. Проверка моделей на вменяемость (sanity check)

Оценим эффективность каждого использованного в исследовании алгоритма с помощью кросс-валидации, для этого импортируем дополнительную функцию cross_val_score.

In [25]:
# импорт функции cross_val_score
from sklearn.model_selection import cross_val_score

Проведем перекрестную проверку для модели **Дерево решений**.

In [26]:
# проведение перекрестной проверки
scores = cross_val_score(model_tree, features_test, target_test, cv=10, n_jobs=-1, scoring="balanced_accuracy")
scores

array([0.67911255, 0.71536797, 0.67911255, 0.79090909, 0.65454545,
       0.68636364, 0.65909091, 0.60681818, 0.69090909, 0.66590909])

In [27]:
# вывод результата
print("Среднее значение accuracy %0.2f со стандартным отклонением %0.2f" % (scores.mean(), scores.std()))

Среднее значение accuracy 0.68 со стандартным отклонением 0.05


Результаты перекрестной проверки неудовлетворительные, в продакшн выводить нельзя.

Проведем перекрестную проверку для модели **Случайный лес**.

In [28]:
# проведение перекрестной проверки
scores = cross_val_score(model_forest, features_test, target_test, cv=10, n_jobs=-1, scoring="balanced_accuracy")
scores

array([0.72781385, 0.70508658, 0.75162338, 0.81590909, 0.64318182,
       0.84318182, 0.68181818, 0.69318182, 0.72954545, 0.78863636])

In [29]:
# вывод результата
print("Среднее значение accuracy %0.2f со стандартным отклонением %0.2f" % (scores.mean(), scores.std()))

Среднее значение accuracy 0.74 со стандартным отклонением 0.06


Результаты перекрестной проверки лучше, чем у модели Дерево решений, однако перед выводом в продакшн требуется доработка.

Проведем перекрестную проверку для модели **Логистическая регрессия**.

In [30]:
# проведение перекрестной проверки
scores = cross_val_score(model_regression, features_test, target_test, cv=10, n_jobs=-1, scoring="balanced_accuracy")
scores

array([0.57142857, 0.5952381 , 0.52380952, 0.65      , 0.5       ,
       0.525     , 0.47727273, 0.55      , 0.62727273, 0.525     ])

In [31]:
# вывод результата
print("Среднее значение accuracy %0.2f со стандартным отклонением %0.2f" % (scores.mean(), scores.std()))

Среднее значение accuracy 0.55 со стандартным отклонением 0.05


Результаты перекрестной проверки неудовлетворительные, в продакшн выводить нельзя.

<div style="border:solid blue 2px; padding: 20px"> 

**Вывод**  

По итогам проведенной перекрестной проверки наилучшие (но не самые оптимальные) результаты показала модель Случайный лес:
    
1. Модель Дерево решений: Среднее значение accuracy 0.68 со стандартным отклонением 0.05
2. Модель Случайный лес: Среднее значение accuracy 0.74 со стандартным отклонением 0.06
3. Модель Логистическая регрессия: Среднее значение accuracy 0.55 со стандартным отклонением 0.05 
  
 
</div>

Для оценки адекватности нашей модели сравним ее метрики качества с **константной моделью**.

Создадим константную модель, прогнозирующую любому объекту класс «1» (is Ultra = True). 

Оценим ее метрики качества на данных обучающей выборки.

In [32]:
# создание константной модели, присваивающей любому объекту класс 1
target_pred_constant = pd.Series(1, index=np.arange(len(target_train)))

# оценка метрик качества константной модели
print('Accuracy:', accuracy_score(target_train, target_pred_constant))
print('Полнота:', recall_score(target_train, target_pred_constant))
print('Точность:', precision_score(target_train, target_pred_constant))
print('F1-мера:',f1_score(target_train, target_pred_constant))

Accuracy: 0.30757261410788383
Полнота: 1.0
Точность: 0.30757261410788383
F1-мера: 0.47044823482744946


Метрики качества константной модели в 3 и более раз хуже, чем у выбранной нами модели Случайный лес с 18 оценщиками, что позволяет сделать вывод об адекватности нашей модели.

## Шаг 6. Общий вывод

<div style="border:solid blue 2px; padding: 20px"> 

**Вывод**  

Целью данного проекта было определение оптимальной модели с точностью ответов не менее 75% для рекомендации клиенту подходящего тарифа на основании использования им тарифных опций.
    
Для решения задач проекта был проведен эксперимент на выборке данных по абонентам тарифов "Ультра" и "Смарт" и осуществлены следующие работы:

1. Шаг 1. Обзор и подготовка данных: 
    
был открыт и изучен файл с выборкой данных, была проведена необходимая предобработка (замена типов данных в колонках calls, minutes, messages, mb_used, is_ultra), а также проверка на наличие пропусков и явных дубликатов.
    
2. Шаг 2. Подготовка к обучению: обучающая, валидационная и тестовая выборки, признаки:
    
для подготовки к обучению данные выборки были разделены на 3 части (60% - обучающая выборка, 20% - валидационная выборка, 20% - тестовая выборка для проверки результатов), для каждой выборки также были назначены признаки и целевой признак.
    
3. Шаг 3. Исследование качества разных моделей с различными вариантами гиперпараметров:
    
на данном шаге исследования с помощью цикла был проведен эксперимент по подбору оптимальных параметров модели для алгоритмов "Дерево решений", "Случайный лес", а также оценены результаты применения алгоритма "Логистическая регрессия";
наилучшие результаты показали модели "Дерево решений" с глубиной 3 (accuracy 0.78) и "Случайный лес" с 18 оценщиками (accuracy 0.79); результаты "Логистической регрессии" оказались наихудшими - accuracy 0.71
    
4. Шаг 4. Проверка качества модели на тестовой выборке:
проверка результатов моделей на тестовой выборке показала, что наилучшие результаты сохраняются у "Случайного леса" и "Дерева решений": accuracy 0.78 и 0.779 соответственно. Точность предсказаний на тестовой выборке снизилась для всех алгоритмов, наихудший результат показала "Логистическая регрессия" - 0.68
    
*Таким образом, по результатам эксперимента в качестве алгоритма для подбора тарифа может быть рекомендована модель "Случайный лес" с 18 оценщиками.*
    
5. Шаг 5. Проверка моделей на вменяемость (sanity check):
    
для проверки на вменяемость была проведена кросс-валидация, которая показала снижение результата по точности предсказания по всем моделям, в том числе, по "Случайному лесу" до 0.74.
    
Из данных кросс-валидации следует, что перед запуском модели "Случайный лес" с 18 оценщиками в продакшн требуется ее донастройка за счет других гиперпараметров для повышения accuracy.
 
 
</div>